### PR TITLE
Implement error::Code and error::Status types

### PR DESCRIPTION
**Pull request by jscherpelz-juniper**
These classes encapsulate common error handling patterns in
exception-free C++. This implementation assumes that errors are
uncommon. In cases where errors are common, other types should be
used.

Note that this implementation differs from the standard Google style
util::Status in that it favors free functions over member functions.
This allows the same API (IsOk(), IsError()) to be used for querying
error::Code and error::Status.

A few basic macros are added to allow documentation and automatic
checking of invariants. When this code should be compiled and run is
left for future discussion.

### CODE REVIEW

File: lib/status.h [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+#include "lib/invariant.h"
+
+namespace error {
+
+// Numeric constants describing common classes of errors.
+//
+// Note that the following helpers exist to check for common conditions:
+// IsOk() and IsError().
+enum class Code {
```
Comment (ccontavalli-j) [inline lib/status.h]: I'm ok with this as is. But: does it make sense to define all those values already? Not convinced it is a particularly good collection.

Does it make sense to make a Code actually be two values, a pointer to a namespace object, and an actual integer value? Good idea, bad idea?

Concern: if the array remains generic, usefulness will be limited. If the array grows over time, than it'll become a single giant array for random error codes in ebplane. Also not good.

I have seen implementations using a namespace object, which were good and very functional (not the one in the status you are familiar with).
Go 2.0 also has a way to wrap an error inside another error, which is *extremely* nice and handy, you can do something like Wrap(ABORTED, othererror).

Eg, you have a timeout that causes a function to fail, you can have a "dataplane programming error" that embeds a "timeout error". Which is very convenient for debugging.

File: lib/BUILD [hunk 2]
```diff
@@ -9,3 +9,31 @@ cc_library(
         "//visibility:public"
     ],
 )
+
```
[the remark this answers is not in the document]
Comment (jscherpelz-juniper) [reply lib/BUILD]: I agree on documentation and discoverability. For all the upsides (users can "extend" types without modifying the implementation) the free function style definitely suffers when it comes to finding functionality (autocomplete rarely helps. Haskell has type-based function lookup to combat this problem) - that's why I made examples inline as often as possible. The other approach that I've found to help is making headers short with limited scope so users don't stop reading before the end.

I'll use a subdirectory and README.md, although I'd like to hold off on the readme until we figure out the details of TRY() and TRY_ASSIGN() if you're willing to put up with a few more macros.

File: lib/status.h (hunk 1 above)
Comment (jscherpelz-juniper) [reply lib/status.h]: I'd actually prefer to use a opaque handle type (actually my next pull request) and allow arbitrary values to be stored in Code, with the single convention being that zero implies success. If you don't mind I'll switch to this approach.

As for error namespaces, I think I view them as a way to decentralize the definition of constants by allocating "ranges" of values in bulk (each namespace implies some combination of high order bits).

The values are just taken from a limited version of util::Status that google open sourced years ago. For me I consider a direct embedding of errno to be most useful, but other applications have different needs.

File: lib/status.cc [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#include "lib/status.h"
+
+namespace error {
+
+std::string_view ToString(const Code code) {
```
[the remark this answers is not in the document]
Comment (jscherpelz-juniper) [reply lib/status.cc]: I'll see what I can get the compiler to do. We need to be robust to arbitrary values since these tend to get sent over the wire.

**Pull request merged**